### ISSUE TITLE
同一数据集下， CPPD在 paddleocr 和 openocr 训练效果有明显差异

### ISSUE THREAD

**Issue by yjc11** [Topdu/OpenOCR#86]
使用两个框架训练时 CPPD 已发现以下 diff：
1. OpenOCR默认的 config 中，Decoder num_layer 默认为 3，paddleocr 中对应的num_layer默认为 2
2. OpenOCR默认的 config 中，默认开启amp，paddleocr 则默认关闭
3. no_weight_decay有差异
3. 两个框架的CPPD 网络有差异，包括 BN 层的 momentum、dropout的 prob等

在对齐 1,2 后，同一数据集下， CPPD在 paddleocr 和 openocr 训练效果有明显差异，如下图所示，同一个验证集上差 3 个点。
>  均为 4 x 128
![Image](https://github.com/user-attachments/assets/d617e2f6-93fc-4de7-9e7a-6ba382178165)

paddleocr config
```yaml
Global:
  use_gpu: True
  epoch_num: 32
  log_smooth_window: 20
  print_batch_step: 20
  save_model_dir: ./output/rec/svtr_cppd_base_ch_from_scratch/
  save_epoch_step: 10
  # evaluation is run every 2000 iterations after the 0th iteration
  eval_batch_step: [0, 2000]
  cal_metric_during_train: False
  pretrained_model: 
  checkpoints:
  save_inference_dir:
  use_visualdl: True
  infer_img: doc/imgs_words_en/word_10.png
  # for data or label process
  character_dict_path: ppocr/utils/ppocr_keys_v1.txt
  max_text_length: 25
  infer_mode: False
  use_space_char: False
  save_res_path: ./output/rec/predicts_svtr_cppd_base_ch_from_scratch_debug.txt



Optimizer:
  name: AdamW
  beta1: 0.9
  beta2: 0.99
  epsilon: 1.e-8
  weight_decay: 0.05
  no_weight_decay_name: norm pos_embed char_node_embed pos_node_embed char_pos_embed vis_pos_embed
  one_dim_param_no_weight_decay: True
  lr:
    name: Cosine
    learning_rate: 0.0005 # 4gpus 128bs
    warmup_epoch: 5

Architecture:
  model_type: rec
  algorithm: CPPD
  Transform:
  Backbone:
    name: SVTRNet
    img_size: [32, 256]
    patch_merging: 'Conv'
    embed_dim: [128, 256, 384]
    depth: [6, 6, 4]
    num_heads: [4, 8, 12]
    mixer: ['Conv','Conv','Conv','Conv','Conv','Conv', 'Conv','Conv', 'Global','Global','Global','Global','Global','Global','Global','Global','Global','Global']
    local_mixer: [[5, 5], [5, 5], [5, 5]]
    last_stage: False
    prenorm: True
  Head:
    name: CPPDHead
    dim: 384
    vis_seq: 128
    ch: &ch True

Loss:
  name: CPPDLoss
  ignore_index: &ignore_index 7000 # must be greater than the number of character classes
  smoothing: True
  sideloss_weight: 1.0

PostProcess:
  name: CPPDLabelDecode

Metric:
  name: RecMetric
  main_indicator: acc

Train:
  dataset:
    name: SimpleDataSet
    data_dir: /workspace/datasets/text_recog_1m/paddle_format_lenless25
    label_file_list:
    - /workspace/datasets/text_recog_1m/paddle_format_lenless25/train.txt
    transforms:
      - DecodeImage: # load image
          img_mode: BGR
          channel_first: False
      - CPPDLabelEncode: # Class handling label
          ignore_index: *ignore_index
          ch: *ch
      - SVTRRecResizeImg:
          image_shape: [3, 32, 256]
          padding: True
      - KeepKeys:
          keep_keys: ['image', 'label', 'label_node', 'label_index', 'length'] # dataloader will return list in this order
  loader:
    shuffle: True
    batch_size_per_card: 128
    drop_last: True
    num_workers: 8

Eval:
  dataset:
    name: SimpleDataSet
    data_dir: /workspace/datasets/docr_rocr_dataset/std_recog_A_v1.2/paddle_format_lenless25
    label_file_list:
    - /workspace/datasets/docr_rocr_dataset/std_recog_A_v1.2/paddle_format_lenless25/labels.txt
    transforms:
      - DecodeImage: # load image
          img_mode: BGR
          channel_first: False
      - CPPDLabelEncode: # Class handling label
          ignore_index: *ignore_index
          ch: *ch
      - SVTRRecResizeImg:
          image_shape: [3, 32, 256]
          padding: True
      - KeepKeys:
          keep_keys: ['image', 'label', 'label_node','label_index','length'] # dataloader will return list in this order
  loader:
    shuffle: False
    drop_last: False
    batch_size_per_card: 128
    num_workers: 2

```

Openocr config
```yaml
Global:
  device: gpu
  epoch_num: 32
  log_smooth_window: 20
  print_batch_step: 20
  output_dir: ./output/rec/ch/svtr_base_cppd_4x128_2/
  save_epoch_step: 10
  eval_batch_step: [0, 2000]
  eval_epoch_step: [0, 1]
  cal_metric_during_train: false
  pretrained_model: null
  checkpoints: null
  use_tensorboard: true
  infer_img: /workspace/datasets/docr_rocr_dataset/std_recog_A_v1.2/paddle_format_lenless25/images
  character_dict_path: ./tools/utils/ppocr_keys_v1.txt
  max_text_length: 25
  use_space_char: false
  save_res_path: ./output/rec/ch/predicts_svtr_base_cppd_4x128_2.txt
  use_amp: false
  project_name: OpenOCR
  run_name: svtr_base_cppd_ch_from_scratch_4x128_2
  distributed: true
Optimizer:
  name: AdamW
  betas:
  - 0.9
  - 0.99
  lr: 0.0005
  weight_decay: 0.05
  filter_bias_and_bn: true

LRScheduler:
  name: CosineAnnealingLR
  warmup_epoch: 5
Architecture:
  model_type: rec
  algorithm: CPPD
  in_channels: 3
  Transform: null
  Encoder:
    name: SVTRNet
    img_size:
    - 32
    - 256
    patch_merging: Conv
    embed_dim:
    - 128
    - 256
    - 384
    depth:
    - 6
    - 6
    - 4
    num_heads:
    - 4
    - 8
    - 12
    mixer:
    - Conv
    - Conv
    - Conv
    - Conv
    - Conv
    - Conv
    - Conv
    - Conv
    - Global
    - Global
    - Global
    - Global
    - Global
    - Global
    - Global
    - Global
    - Global
    - Global
    local_mixer:
    - - 5
      - 5
    - - 5
      - 5
    - - 5
      - 5
    last_stage: false
    prenorm: true
  Decoder:
    name: CPPDDecoder
    vis_seq: 128
    num_layer: 2
    pos_len: false
    rec_layer: 1
    dim: 384
    ch: true
Loss:
  name: CPPDLoss
  ignore_index: 7000
  smoothing: true
  pos_len: false
  sideloss_weight: 1.0
PostProcess:
  name: CPPDLabelDecode
  character_dict_path: ./tools/utils/ppocr_keys_v1.txt
  use_space_char: false
Metric:
  name: RecMetric
  main_indicator: acc
Train:
  dataset:
    name: SimpleDataSet
    data_dir: /workspace/datasets/text_recog_1m/paddle_format_lenless25
    label_file_list:
    - /workspace/datasets/text_recog_1m/paddle_format_lenless25/train.txt
    transforms:
    - DecodeImage:
        img_mode: BGR
        channel_first: false
    - CPPDLabelEncode:
        pos_len: false
        ch: true
        ignore_index: 7000
        character_dict_path: ./tools/utils/ppocr_keys_v1.txt
        use_space_char: false
        max_text_length: 25
    - SVTRResize:
        image_shape:
        - 3
        - 32
        - 256
        padding: true
    - KeepKeys:
        keep_keys:
        - image
        - label
        - label_node
        - label_index
        - length
  loader:
    shuffle: true
    batch_size_per_card: 128
    drop_last: true
    num_workers: 8
Eval:
  dataset:
    name: SimpleDataSet
    data_dir: /workspace/datasets/docr_rocr_dataset/std_recog_A_v1.2/paddle_format_lenless25
    label_file_list:
    - /workspace/datasets/docr_rocr_dataset/std_recog_A_v1.2/paddle_format_lenless25/labels.txt
    transforms:
    - DecodeImage:
        img_mode: BGR
        channel_first: false
    - CPPDLabelEncode:
        pos_len: false
        ch: true
        ignore_index: 7000
        character_dict_path: ./tools/utils/ppocr_keys_v1.txt
        use_space_char: false
        max_text_length: 25
    - SVTRResize:
        image_shape:
        - 3
        - 32
        - 256
        padding: true
    - KeepKeys:
        keep_keys:
        - image
        - label
        - label_node
        - length
  loader:
    shuffle: false
    drop_last: false
    batch_size_per_card: 256
    num_workers: 4
filename: svtr_base_cppd_ch
config: configs/rec/cppd/svtr_base_cppd_ch.yml
local_rank: null
eval: true
```

想知道 OpenOCR 中的 CPPD 是否有跟 PaddleOCR 中的 CPPD 严谨对齐，如果确实有差异，作者是否能更新对齐的代码，感谢！

**Comment by Topdu** [comment]
非常感谢发现这个潜在的问题。然而，在我们进行的对比实验中未发现这种情况。
我们保证了模型结构是一致的，最终精度为: openocr (w/ amp) 72.82, openocr (w/o amp) 73.23, paddleocr (w/o amp) 73.28。
实验数据为bctr的scene子集，训练100epoch。

猜测原因可能是32epoch未将模型训练收敛，造成结果差距比较大。

**Comment by yjc11** [comment]
感谢回复，cppd 是非常优秀的工作。目前本人已逐层将网络对齐，目前将 paddle 权重加载到 torch 网络上，推理结果能完全对齐（6k 个样本的推理结果仅有一个字的差异），TRAIN 模式下，loss 也能完全对齐。对齐过程中发现的网络部分差异如下：
1.  BN 的 momentum不同是由于 paddle 和 torch 的公式有差异，实则是等价的。
2. CPPDDecoder的edge_decoder相关实现有差异, openocr为 DecoderLayer， paddleocr 为EdgeDecoderLayer。forward 方法也有差异。
3. svtrnet ConvBNLayer的 bias有差异，openocr 为 False，paddleocr 为 True

其余差异：
1. weight decay。Openocr为 
```
{'char_pos_embed', 'vis_pos_embed', 'char_node_embed','pos_node_embed' , 'pos_embed', 'sub_sample1', 'sub_sample2', 'sub_sample3'}
```
paddleocr为
```
norm pos_embed char_node_embed pos_node_embed char_pos_embed vis_pos_embed
```

如果作者需要我可以提个 pr。
希望此 issue 能帮助遇到类似问题的朋友们。

**Comment by Topdu** [comment]
非常感谢您细致的检查与反馈！针对您提到的几个问题，我们进行了进一步的分析和总结，具体如下：

1. **关于CPPDDecoder中两种`edge_decoder`实现的差异**  
   经过确认，这两种实现方式仅在代码结构上存在差异，其核心计算逻辑是一致的。因此，OpenOCR和PaddleOCR的权重在理论上是可以互相转换的。

2. **关于`bias`参数的差异**  
   提到的`bias`差异问题，是由于PyTorch和PaddlePaddle框架在默认参数设置上的不同所导致的。

3. **关于`no_weight_decay_name`的差异**  
   针对这一部分的差异，我们计划开展一组对比实验，以验证其对模型性能的具体影响。实验结果将及时反馈给您。

4. **关于PR和代码共享**  
   我们非常欢迎并期待您的PR贡献！同时，如果您能够共享PaddleOCR到OpenOCR的转换代码，这将极大地促进两个框架之间的兼容性和协作。我们对此表示由衷的感谢，期待与您共同推动项目的进一步发展。

**Comment by yjc11** [comment]
感谢回复，我非常愿意共享我的代码，近期对齐工作完成后，我会整理相关代码提个 pr。另外今天在对齐训练过程中还发现了两个 diff：
1. 两个框架lr schedule实现有差异，墨绿色为 paddleocr，青色为openocr，差异如下图所示

<img width="251" alt="Image" src="https://github.com/user-attachments/assets/c1c9dd65-b987-4bb5-bec9-b3e6a244967e" />

2. 默认配置下 AdamW 的 betas 有差异，paddleocr 为 [0.9,0.99], openocr 为[0.9, 0.999]

**Comment by Topdu** [comment]
关于no_weight_decay_name的差异，我们将weight decay 对齐后，即将'sub_sample1', 'sub_sample2', 'sub_sample3'从no_weight_decay_name中去除，openocr (w/o amp) 的精度为73.19。

学习率的差异我们将进一步对齐，感谢您提出这个问题！！

**Comment by zhuzhiqiang148** [comment]
I trained the CPPD algorithm using both PPOCR and OpenOCR frameworks, but it fails to converge on either public or private datasets. All training parameters were kept at their default settings. I would like to ask what might be the reason for this issue.

**Comment by zhuzhiqiang148** [comment]
Global:
  use_gpu: True
  epoch_num: 500
  log_smooth_window: 20
  print_batch_step: 10
  save_model_dir: ./output/rec/svtr_cppd_base/
  save_epoch_step: 10 #1
  # evaluation is run every 2000 iterations after the 0th iteration
  eval_batch_step: [0, 2000]
  cal_metric_during_train: True
  pretrained_model:
  checkpoints:
  save_inference_dir:
  use_visualdl: False
  infer_img: doc/imgs_words_en/word_10.png
  # for data or label process
  character_dict_path: /home/ubuntu/PaddleOCR-main/ppocr/utils/cepen.txt
  character_type: en
  max_text_length: 25
  infer_mode: False
  use_space_char: False
  save_res_path: ./output/rec/predicts_svtr_cppd_base.txt


Optimizer:
  name: AdamW
  beta1: 0.9
  beta2: 0.99
  epsilon: 1.e-8
  weight_decay: 0.05
  no_weight_decay_name: norm pos_embed char_node_embed pos_node_embed char_pos_embed vis_pos_embed
  one_dim_param_no_weight_decay: True
  lr:
    name: Cosine
    learning_rate: 0.0005 # 4gpus 256bs
    warmup_epoch: 2

Architecture:
  model_type: rec
  algorithm: CPPD
  Transform:
  Backbone:
    name: SVTRNet
    img_size: [32, 100]
    patch_merging: 'Conv'
    embed_dim: [128, 256, 384]
    depth: [6, 6, 4]
    num_heads: [4, 8, 12]
    mixer: ['Conv','Conv','Conv','Conv','Conv','Conv', 'Conv','Conv', 'Global','Global','Global','Global','Global','Global','Global','Global','Global','Global']
    local_mixer: [[5, 5], [5, 5], [5, 5]]
    last_stage: False
    prenorm: True
  Head:
    name: CPPDHead
    dim: 384
    vis_seq: 50
    num_layer: 3

Loss:
  name: CPPDLoss
  ignore_index: &ignore_index 100 # must be greater than the number of character classes
  smoothing: True
  sideloss_weight: 1.0

PostProcess:
  name: CPPDLabelDecode

Metric:
  name: RecMetric
  main_indicator: acc

Train:
  dataset:
    name: LMDBDataSet
    data_dir: /home/ubuntu/PaddleOCR-main/datasets/data_cepen_3w/training/
    transforms:
      - DecodeImage: # load image
          img_mode: BGR
          channel_first: False
      - CPPDLabelEncode: # Class handling label
          ignore_index: *ignore_index
      - SVTRRecResizeImg:
          image_shape: [3, 32, 100]
          padding: False
      - KeepKeys:
          keep_keys: ['image', 'label', 'label_node', 'length'] # dataloader will return list in this order
  loader:
    shuffle: True
    batch_size_per_card: 256
    drop_last: True
    num_workers: 8

Eval:
  dataset:
    name: LMDBDataSet
    data_dir: /home/ubuntu/PaddleOCR-main/datasets/data_cepen_3w/validation/
    transforms:
      - DecodeImage: # load image
          img_mode: BGR
          channel_first: False
      - CPPDLabelEncode: # Class handling label
          ignore_index: *ignore_index
      - SVTRRecResizeImg:
          image_shape: [3, 32, 100]
          padding: False
      - KeepKeys:
          keep_keys: ['image', 'label', 'label_node','length'] # dataloader will return list in this order
  loader:
    shuffle: False
    drop_last: False
    batch_size_per_card: 256
    num_workers: 2
